osgp/platform/osgp-adapter-domain-core/src/test/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementServiceTest.java

```diff
@@ -177,4 +209,184 @@ public void testTryToAddFirmwareVersionToHistoryWhenFileIsNotAvailable() throws
+                olderPendingFirmwareUpdate2, olderPendingFirmwareUpdate3);
+
+        /*
+         * Check that a get firmware version message is sent for the most recent
+         * pending firmware update.
+         */
+        this.assertFirmwareVersionRequested(organisation.getOrganisationIdentification(), deviceIdentification,
+                correlationUidMostRecentPendingFirmwareUpdate);
+    }
+
+    private void assertFirmwareVersionRequested(final String organisationIdentification,
+            final String deviceIdentification, final String correlationUid) {
+
+        final ArgumentCaptor<RequestMessage> requestMessageCaptor = ArgumentCaptor.forClass(RequestMessage.class);
+        verify(this.osgpCoreRequestMessageSender).sendWithDelay(requestMessageCaptor.capture(),
+                eq(MessageType.GET_FIRMWARE_VERSION.name()), anyInt(), any(), any());
+
+        final RequestMessage actualRequestMessage = requestMessageCaptor.getValue();
+        assertThat(actualRequestMessage.getCorrelationUid()).isEqualTo(correlationUid);
+        assertThat(actualRequestMessage.getDeviceIdentification()).isEqualTo(deviceIdentification);
+        assertThat(actualRequestMessage.getOrganisationIdentification()).isEqualTo(organisationIdentification);
+    }
+
+    private SsldPendingFirmwareUpdate anSsldPendingFirmwareUpdate(final Long id, final Date creationTime,
+            final String deviceIdentification, final String correlationUid) {
+
+        final FirmwareModuleType firmwareModuleType = FirmwareModuleType.FUNCTIONAL;
+        final String firmwareVersion = "test-version";
+        final String organisationIdentification = "test-org";
+        final SsldPendingFirmwareUpdate ssldPendingFirmwareUpdate = new SsldPendingFirmwareUpdate(deviceIdentification,
+                firmwareModuleType, firmwareVersion, organisationIdentification, correlationUid);
+        ReflectionTestUtils.setField(ssldPendingFirmwareUpdate, "id", id, Long.class);
+        ReflectionTestUtils.setField(ssldPendingFirmwareUpdate, "creationTime", creationTime, Date.class);
+        return ssldPendingFirmwareUpdate;
+    }
+
+    void checkSsldPendingFirmwareUpdateReturnsFalseIfThereAreNoPendingUpdates() {
+        final String organisationIdentification = "test-org";
+        final String deviceIdentification = "device-identification";
+        final String correlationUid = "correlation-uid-no-pending-updates";
+        final CorrelationIds ids = new CorrelationIds(organisationIdentification, deviceIdentification, correlationUid);
+        final List<FirmwareVersion> firmwareVersions = Collections
+                .singletonList(new FirmwareVersion(FirmwareModuleType.FUNCTIONAL, VERSION_3));
+        when(this.ssldPendingFirmwareUpdateRepository.findByDeviceIdentification(deviceIdentification))
+                .thenReturn(Collections.emptyList());
+
+        assertThat(this.firmwareManagementService.checkSsldPendingFirmwareUpdate(ids, firmwareVersions)).isFalse();
```

Comment:
```suggestion
        // act
        boolean hasPendingFirmwareUpdate = this.firmwareManagementService.checkSsldPendingFirmwareUpdate(ids, firmwareVersions);

        // assert
        assertThat(hasPendingFirmwareUpdate).isFalse();
```

Reply:
In my opinion the suggestion would lead to a cleaner separation of steps in arrange-act-assert approach. 
(Same applies to other tests below)
Although this is probably a matter of opinion...

Reply:
While arrange-act-assert makes for a good test approach, I quite strongly dislike the addition of comments indicating these separate stages.

I don't think that stating the obvious is a good reason to add a comment. Comments should rather explain things that could use some explanation.

Something like the following looks like needless duplication:
// assert
assertThat ...
This falls, I think, in the same category as comments like:
x = y +1; // make x one more than y

In my taste, adding the comments makes the tests look like a beginners tutorial on testing, instead of real world project test code. While I am all for clear and understandable code, I think there is a line between that and treating every reader as a programming novice, which is, in my opinion, crossed adding the comments as you propose.

As you said, apparently opinions can vary...

Reply:
As discussed, the suggestion was not as much about adding the comments, but more about splitting up the act-assert steps that are currently on one line to have a cleaner separation between the arrange-act-assert steps. 
Still matter of opinion, no deal-breaker...

Reply:
OK, I did not read that carefully enough the first time,
I don't think it is necessarily better to always have the code acting and asserting on separate lines.
Looking at this particular case again, I tend to agree with your suggestion, but for the additional reason that the name of the return value makes the meaning of the return value a bit more obvious.
I'll also try to put the w in firmware this time, so we won't need a third round of changes! ;)